File: 2868-continuous-subarrays/2868-continuous-subarrays.java
class Solution {
    public long continuousSubarrays (int [] nums) {
        int n = nums.length;
        int j = 0;
        long count = 0;

        Map <Integer, Integer> map = new HashMap <> ();

        for (int i = 0; i < n; i++) {
            map.put (nums [i], map.getOrDefault (nums [i], 0) + 1);

            while ((i - j + 1) > getCount (nums [i], map)) {
                map.put (nums [j], map.get (nums [j]) - 1);
                j++;
            }
            count += (i - j + 1);
        }
        return count;
    }

    private int getCount (int num, Map <Integer, Integer> map) {
        return map.getOrDefault (num, 0) + map.getOrDefault (num - 1, 0) +
            map.getOrDefault (num + 1, 0) + map.getOrDefault (num - 2, 0) +
            map.getOrDefault (num + 2, 0);
    }
}

File: 2868-continuous-subarrays/2868-continuous-subarrays.c
long long continuousSubarrays (int* nums, int numsSize){

    long long count = 0;
    int left = 0, right = 0;

    int maxDeque [numsSize], minDeque [numsSize];
    int maxStart = 0, maxEnd = -1, minStart = 0, minEnd = -1;

    while (right < numsSize) {
        int current = nums [right];

        while (maxEnd >= maxStart && maxDeque [maxEnd] < current) {
            maxEnd--;
        }
        maxDeque [++maxEnd] = current;

        while (minEnd >= minStart && minDeque [minEnd] > current) {
            minEnd--;
        }
        minDeque [++minEnd] = current;

        while (maxDeque [maxStart] - minDeque [minStart] > 2) {
            if (nums [left] == maxDeque [maxStart]) {
                maxStart++;
            }
            if (nums [left] == minDeque [minStart]) {
                minStart++;
            }
            left++;
        }

        count += (right - left + 1);
        right++;
    }

    return count;
}